# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1746907996 -7200
# Parent  c7c35cdf8835b65c18d22b5dc9fab214852776c5
NNN27 (128-only): Fix blank thread pane after repair folder.

This is a regression from https://bugzilla.mozilla.org/show_bug.cgi?id=1933104 which showed up on 128.6.0 when that bug got uplifted.
Curiously enough, the issue was fixed in this range on trunk at TB 139 here.
https://hg.mozilla.org/comm-central/pushloghtml?fromchange=63eca98887dd4e306fcb8530f900e3ce91d1059e&tochange=07dce77762205a4207bb4a7d72ab4c1aed176a7b
It's not clear what fixed it, certainly some M-C change. Before the fix, `requestAnimationFrame()` in tree-view.mjs only returned when the modal panel
was dismissed.

So here's an 128-only fix.

diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -3845,8 +3845,15 @@ var folderPane = {
     }
 
     // The following notification causes all DBViewWrappers that include
     // this folder to rebuild their views.
+
+    // Totally ugly hack to force drawing the rows since `requestAnimationFrame()` over in
+    // tree-view.mjs only returns when the modal panel is dismissed. That got fixed on M-C,
+    // but to late for 128.
+    if (gFolder == folder) {
+      threadTree.forceDraw = true;
+    }
     MailServices.mfn.notifyFolderReindexTriggered(folder);
 
     folder.msgDatabase.summaryValid = false;
     const isIMAP = folder.server.type == "imap";
@@ -3875,8 +3882,15 @@ var folderPane = {
       folder.QueryInterface(
         Ci.nsIMsgImapMailFolder
       ).suppressNotification = true;
     }
+    window.addEventListener(
+      "allMessagesLoaded",
+      () => {
+        threadTree.forceDraw = false;
+      },
+      { once: true }
+    );
     folder.updateFolder(top.msgWindow);
   },
 
   /**
diff --git a/mail/base/content/widgets/tree-view.mjs b/mail/base/content/widgets/tree-view.mjs
--- a/mail/base/content/widgets/tree-view.mjs
+++ b/mail/base/content/widgets/tree-view.mjs
@@ -2729,8 +2729,15 @@ export class TreeViewTableRow extends HT
 
     // Wait before filling the row. This setter could be called many times
     // before it even appears on the screen, and calling (potentially very
     // expensive) code each time would be a waste.
+    
+    // Totally ugly hack used during folder repair.
+    if (this.parentNode.parentNode?.parentNode?.forceDraw) {
+      this._fillRow();
+      return;
+    }
+
     if (!this.#animationFrame) {
       this.#animationFrame = requestAnimationFrame(() => {
         this.#animationFrame = null;
         // The row may no longer be attached to the tree. Don't waste time
